### PR TITLE
Add Llama 4 (Scout/Maverick) text-decoder model

### PR DESCRIPTION
**Pull request by astorise**
## Summary

Adds a Candle implementation of the Llama 4 text-only decoder (Scout / Maverick), Meta's MoE-based LLM family:

- iRoPE: alternating RoPE/NoPE attention layers, configurable via `no_rope_layers` or `no_rope_layer_interval`
- Parameterless QK-norm (L2 normalization), applied only on RoPE layers
- Attention temperature tuning on NoPE layers
- Interleaved-pair RoPE (`rope_i`) matching the reference implementation
- Top-k sigmoid-gated MoE block with an always-on shared expert, using the real checkpoint's batched expert parameter layout (`experts.gate_up_proj` / `experts.down_proj`, applied with a per-expert matmul rather than per-expert `nn.Linear` sub-modules)
- Dense-vs-MoE layer selection via `interleave_moe_layer_step`

Includes a new `llama4` example (`candle-examples/examples/llama4`) mirroring the existing `deepseekv2`/`llama` examples, plus a README.

## Testing

- Unit test (`forward_smoke_test`) exercising prefill + KV-cache decode with synthetic weights.
- New `#[ignore]`d end-to-end integration test (`candle-examples/tests/llama4_integration.rs`) that downloads `yujiepan/llama-4-tiny-random` — a randomly-initialized checkpoint using the real Llama 4 architecture — and runs a real prefill + decode step through the production model code:

  ```
  cargo test -p candle-examples --release --test llama4_integration -- --ignored
  ```

  This caught three real-checkpoint incompatibilities that synthetic-weight unit tests couldn't:
  - `attn_temperature_tuning` can be serialized as a raw integer rather than a JSON boolean in some configs
  - MoE expert weights are stored as batched parameter tensors, not per-expert named sub-weights
  - the shared expert uses `intermediate_size`, not `intermediate_size_mlp`
- `cargo fmt` / `cargo clippy` clean on `candle-transformers` and `candle-examples`.
- Manually ran `cargo run --example llama4 -- --model-id yujiepan/llama-4-tiny-random --prompt "Hello" --sample-len 5` end-to-end.

The real Scout (109B total / 17B active) and Maverick (400B total / 17B active) checkpoints are gated on the Hub and too large to exercise in CI; the tiny-random checkpoint validates the loading/shape/architecture path against the real Llama4ForConditionalGeneration weight layout instead.

## Test plan

- [x] `cargo test -p candle-transformers --lib llama4`
- [x] `cargo test -p candle-examples --release --test llama4_integration -- --ignored`
- [x] `cargo fmt --check` / `cargo clippy`
- [x] `cargo run --example llama4 -- --prompt "..." --sample-len ...` manually

### CODE REVIEW

**Comment by astorise** [comment]
« Closing — too broad for now. I'll revisit as a smaller, isolated change if there's interest. »
